lib/Target/CBackend/CBackend.cpp

```diff
@@ -906,8 +906,14 @@ bool CWriter::printConstantString(Constant *C, enum OperandContext Context) {
 
   StringRef Bytes = CDS->getAsString();
 
-  // Do not include the last character, which we know is null
-  for (unsigned i = 0, e = Bytes.size() - 1; i < e; ++i) {
+  unsigned length = Bytes.size();
+  // We can skip the last character only if it is an implied null.
+  // Beware: C does not force character (i.e. (u)int8_t here) arrays to have a
+  // null terminator, but if the length is not specified it will imply one!
+  if (length >= 1 && Bytes[length - 1] == '\0')
```

Comment:
I think LLVM always specifies the length of the array, so C would actually specify all trailing bytes are null, so this would also probably be valid (though unnecessary):
```suggestion
  while (length > 0 && Bytes[length - 1] == '\0')
```

Reply:
That may be right, but I think it might be surprising if all the trailing null bytes were dropped.